compiler/sem/semexprs.nim

```diff
@@ -2370,10 +2370,17 @@ proc semProcBody(c: PContext, n: PNode): PNode =
       # are not expressions:
       fixNilType(c, result)
     else:
-      var a = newNodeI(nkAsgn, n.info, 2)
-      a[0] = newSymNode(c.p.resultSym)
-      a[1] = result
-      result = semAsgn(c, a)
+      if result.kind == nkStmtListExpr:
+        # in order to preserve doc comments, apply the return to the last
+        # statement in the list, not to the whole list
+        result.transitionSonsKind(nkStmtList)
+        result.typ = nil
+        let last = semReturn(c, newTreeI(nkReturnStmt, n.info, result[^1]))
```

Comment:
I can't recall, but is it possible to end a statement list expression with a doc comment (and an expression being the last non-doc comment node)?

If that's the case then we'll need to search for the last expression, otherwise this code should be fine.

Reply:
> I can't recall, but is it possible to end a statement list expression with a doc comment (and an expression being the last non-doc comment node)?

It's allowed by the syntax, but not the semantics. `semStmtList` treats `nkCommentStmt` like real statements (except when reporting `unreachable code` warnings), and thus they can never be the last item in a valid `nkStmtListExpr`.

The `return` wrapping takes place *after* the body expression was typed, meaning that only valid `nkStmtListExpr` reach here.